### ISSUE TITLE
binMean1D() and binApply1D() handling of on-breakpoint data is not R-ish

### ISSUE THREAD

**Issue by dankelley** [dankelley/oce#2104]
This came up in #2103 and, in fact, I am going to copy comments in that issue to here (deleting them at #2103) in the next few minutes.  I am often guilty of issue-drift but I want this topic, which is likely to cause user-visible changes, to be well documented.  The work will take a few days.

**Comment by dankelley** [comment]
NB. this comment is an edited version of a comment that was originally at #2103.  I will delete the original comment as soon as I finalize this one.

The results from `binMean1D()` and `binApply1D()` can differ at an endpoint (and, also at interior points).  @richardsc and I have talked about this f2f today, so I won't spend much time trying to find a clear way of explaining things here.  The basic question is whether we want to look for things in intervals that are open-at-left and closed-at-right or vice versa.

We do `binMean1D()` in R that calls C++ (more below) and I think I want to change the C++ function.

This has to do with the default vallues of the `include.lowest` and `right` arguments of `cut()`.

I think `binMean1D()` ought to follow the lead of `cut()`, because the latter is likely used by thousands of people everyday and is likely decades old, so it must be "best practice" or at least "most-expected practice".

**Comment by dankelley** [comment]
NB. this comment is an edited version of a comment that was originally at https://github.com/dankelley/oce/issues/2103. I will delete the original comment as soon as I finalize this one.


In a nutshell:

* `binMean1D()` uses C++ `std::lower_bound(first, last)`.  Q: should we use `std::upper_bound()` instead?

* `binApply1D()` uses R `cut()`, the defaults for which amount to `(first, last]`, that is, discards points that match `first` but keeps points if they equal `last`.

(I may have this wrong ... just typing in another meeting.). But the main point is that I want to bring the two things into line.

**Comment by dankelley** [comment]
NB. this comment is an edited version of a comment that was originally at https://github.com/dankelley/oce/issues/2103. I will delete the original comment as soon as I finalize this one.


I'm looking at the C++ functions to see how to make `binMean1D()` match `binApply1D()` in cases in which some `x` values fall exactly on `breaks` boundaries.  Refs (so I don't have to bookmark):
* https://en.cppreference.com/w/cpp/algorithm/upper_bound (what might be better)
* https://en.cppreference.com/w/cpp/algorithm/lower_bound (what we have now)

**Comment by dankelley** [comment]
NB. this comment is an edited version of a comment that was originally at https://github.com/dankelley/oce/issues/2103. I will delete the original comment as soon as I finalize this one.

More notes to self: I translate the docs as below.  (The docs are written in terms of things that can be located by "iterator" object.  The iterator concept is subtle and I won't explain it here.  C++ can deal with a lot more than simple vectors of numbers etc.)

* https://en.cppreference.com/w/cpp/algorithm/lower_bound 

Smallest `i` such that `x <= breaks[i]`, or `length(breaks)` if no such `i` is in `seq_along(breaks)`.

* https://en.cppreference.com/w/cpp/algorithm/upper_bound 

Smallest `i` such that `x < breaks[i]`, or `length(breaks)` if no such `i` is in `seq_along(breaks)`.

**Comment by dankelley** [comment]
NB. this comment is an edited version of a comment that was originally at https://github.com/dankelley/oce/issues/2103. I will delete the original comment as soon as I finalize this one.

Over the next few days I'll clear this up.  I'm just taking notes here.  I won't bother explaining details since that takes as much time as reading and coding.  The main thing is that `cut()` and `binCount1D()` do different things for points on boundaries.  (Which is preferable is sort of an open question to me.

```R
> cut(0:10, seq(0, 10, 2))
 [1] <NA>   (0,2]  (0,2]  (2,4]  (2,4]  (4,6]  (4,6]  (6,8]  (6,8]  (8,10]
[11] (8,10]
```

notice the NA at the start.  A simpler case is

```R
> cut(0, seq(0, 10, 2))
[1] <NA>
Levels: (0,2] (2,4] (4,6] (6,8] (8,10]
```
Now, notice what happens if I give an arg
```R
> cut(0:10, seq(0, 10, 2), include.lowest=TRUE)
 [1] [0,2]  [0,2]  [0,2]  (2,4]  (2,4]  (4,6]  (4,6]  (6,8]  (6,8]  (8,10]
[11] (8,10]
Levels: [0,2] (2,4] (4,6] (6,8] (8,10]
```
and for the simpler case
```R
> cut(0, seq(0, 10, 2), include.lowest=TRUE)
[1] [0,2]
Levels: [0,2] (2,4] (4,6] (6,8] (8,10]
```


These illustrate the "R way".  However, a sort-of-similar oce function does as follows (at the moment):

```R
> binCount1D(0:20, seq(0, 20, 5))
$xbreaks
[1]  0  5 10 15 20

$xmids
[1]  2.5  7.5 12.5 17.5

$number
[1] 5 5 5 5
```

and 
```R
> binCount1D(0, seq(0, 20, 5))
$xbreaks
[1]  0  5 10 15 20

$xmids
[1]  2.5  7.5 12.5 17.5

$number
[1] 1 0 0 0
```

**Comment by dankelley** [comment]
The `binMean1D()` case was solved yesterday (see #2112) so I removed it from the title and will link to this in #2113.

**Comment by dankelley** [comment]
I'm closing this now, because the two issues are covered in #2112 and #2114
